js/ui/codingManager.js

```diff
@@ -0,0 +1,638 @@
+            session.actorApp.meta_window.disconnect(session.positionChangedIdApp);
+            session.positionChangedIdApp = 0;
+        }
+        if (session.sizeChangedIdApp !== 0) {
+            session.actorApp.meta_window.disconnect(session.sizeChangedIdApp);
+            session.sizeChangedIdApp = 0;
+        }
+        if (session.windowsRestackedId !== 0) {
+            Main.overview.disconnect(session.windowsRestackedId);
+            session.windowsRestackedId = 0;
+        }
+        if (session.windowMinimizedId !== 0) {
+            global.window_manager.disconnect(session.windowMinimizedId);
+            session.windowMinimizedId = 0;
+        }
+        if (session.windowUnminimizedId !== 0) {
+            global.window_manager.disconnect(session.windowUnminimizedId);
+            session.windowUnminimizedId = 0;
+        }
+
+        Main.layoutManager.removeChrome(session.buttonApp);
+        session.buttonApp.destroy();
+        session.actorApp = null;
+
+        // Builder window still open, keep it open
+        // but remove the coding session specific parts
+        if (session.actorBuilder) {
+            this._clearBuilderSession(session);
+
+            session.actorBuilder.meta_window.activate(global.get_current_time());
+            session.actorBuilder.show();
+        } else {
+            this._removeSession(session);
+        }
+    },
+
+    _clearBuilderSession: function(session) {
+        if (session.positionChangedIdBuilder !== 0) {
+            session.actorBuilder.meta_window.disconnect(session.positionChangedIdBuilder);
+            session.positionChangedIdBuilder = 0;
+        }
+        if (session.sizeChangedIdBuilder !== 0) {
+            session.actorBuilder.meta_window.disconnect(session.sizeChangedIdBuilder);
+            session.sizeChangedIdBuilder = 0;
+        }
+        if (session.buttonBuilder) {
+            Main.layoutManager.removeChrome(session.buttonBuilder);
+            session.buttonBuilder.destroy();
+            session.buttonBuilder = null;
+        }
+    },
+
+    _removeSwitcherToApp: function(actorBuilder) {
+        let session = this._getSession(actorBuilder);
+        if (!session)
+            return;
+
+        this._clearBuilderSession(session);
+        session.actorBuilder = null;
+
+        if (session.actorApp) {
+            session.actorApp.meta_window.activate(global.get_current_time());
+            session.actorApp.show();
+            session.buttonApp.show();
+        } else {
+            this._removeSession(session);
+        }
+    },
+
+    _switchToBuilder: function(actor, event, session) {
+        if (!session.actorBuilder) {
+            let tracker = Shell.WindowTracker.get_default();
+            tracker.track_coding_app_window(session.actorApp.meta_window);
+            this._watchdogId = Mainloop.timeout_add(WATCHDOG_TIME,
+                                                    Lang.bind(this, this._watchdogTimeout));
```

Comment:
You could also save the watchdog in the session, but for now we can also keep it here